lib/parser.js

```diff
@@ -1,41 +1,186 @@
 'use strict';
 
-const HANDSHAKE_LENGTH = 36;
-const PARCEL_LENGTH = 16;
-const CHUNK_LENGTH = 12;
+const constants = {
+  PROTOCOL_VERSION: 1,
+  HANDSHAKE_LENGTH: 36,
+  PARCEL_LENGTH: 16,
+  CHUNK_LENGTH: 12,
+  STRUCT_PARCEL: 0,
+  STRUCT_CHUNK: 1,
+  PARCEL_PING: 0,
+  PARCEL_PONG: 1,
+  PARCEL_CALL: 2,
+  PARCEL_CALLBACK: 3,
+  PARCEL_EVENT: 4,
+  PARCEL_STREAM: 5
+};
+
+const partPayload = (
+  // Turn data into an array of payloads
+  string, // string, all payload
+  size = 2048 // number, max chunk payload length
+  // Returns: buffer
+) => {
+  let chunkId = 1;
+  if (!string.length)
+    return [{
+      chunkId,
+      payload: Buffer.alloc(0),
+      length: 0,
+    }];
 
-const parcel = (buffer) => {
+  let payload;
+  const arrayOfPayloads = [];
+  while (string.length > 0) {
+    payload = string.substring(0, size);
+    string = string.slice(size);
+    arrayOfPayloads.push({
+      chunkId: chunkId++,
+      payload: Buffer.from(payload),
+      length: payload.length
+    });
+  }
+
+  return arrayOfPayloads;
+};
+
+const readParcel = (
   // Read parcel from buffer
-  // Return parcel object
-  const rest = Buffer.from(buffer, 1, PARCEL_LENGTH);
-  return rest;
+  buffer // buffer, containing parcel
+  // Returns: object
+) => {
+  const parcel = {
+    structType: constants.STRUCT_PARCEL,
+    parcelId: buffer.readIntLE(1, 4),
+    parcelType: buffer.readIntLE(5, 1),
+    compression: buffer.readIntLE(6, 1),
+    encoding: buffer.readIntLE(7, 1),
+    length: buffer.readIntLE(8, 6),
+  };
+  return parcel;
+};
+
+const parcel = (
+  // Encode parcel structure to Buffer
+  parcel // object, { parcelId, parcelType, compression, encoding, length }
+  //  parcelId - number, id of parcel
+  //  parcelType - number, type of parcel
+  //  compression - number, compression type
+  //  encoding - number, encoding type
+  //  length - number, length of all parcel chunks payload
+  // Returns: buffer
+) => {
+  const buffer = Buffer.alloc(constants.PARCEL_LENGTH);
+  buffer.writeIntLE(constants.STRUCT_PARCEL, 1);
+  buffer.writeIntLE(parcel.parcelId, 1, 4);
+  buffer.writeIntLE(parcel.parcelType, 5, 1);
+  buffer.writeIntLE(parcel.compression, 6, 1);
+  buffer.writeIntLE(parcel.encoding, 7, 1);
+  buffer.writeIntLE(parcel.length, 8, 6);
+  return buffer;
 };
 
-const chunk = (buffer) => {
-  // Read chunkk from buffer
-  // Return chunk object
-  const rest = Buffer.from(buffer, 1, CHUNK_LENGTH);
-  return rest;
+const readChunk = (
+  // Read chunk from buffer
+  buffer // buffer, containing chunk
+  // Returns: object
+) => {
+  const length = buffer.readIntLE(10, 2);
```

Comment:
I have some doubts, because we use it when slicing buffer to get a payload, and we can't write it like this
```
const chunk = {
  ...
  length: buffer.readIntLE(10, 2),
  payload: buffer.slice(12, parser.constants.CHUNK_LENGTH + chunk.length)
}
```
because this throws an error.